src/charm.py

```diff
@@ -0,0 +1,469 @@
+        )
+
+        self.framework.observe(self.on[CERTIFICATES].relation_joined, self._tls_relation_joined)
+        self.framework.observe(
+            self.certificates.on.certificate_available, self._on_certificate_available
+        )
+
+        # actions
+        self.framework.observe(
+            getattr(self.on, "start_process_action"), self._start_process_action
+        )
+        self.framework.observe(getattr(self.on, "stop_process_action"), self._stop_process_action)
+
+    def _tls_relation_joined(self, event: EventBase) -> None:
+        """Handle the tls relation joined event."""
+        if not self.peer_relation.app_data:
+            event.defer()
+            return
+
+        self.peer_relation.set_private_key(generate_private_key().decode("utf-8"))
+        self._request_certificate()
+
+    def _request_certificate(self):
+        """Request the certificate."""
+        if not self.peer_relation.app_data:
+            return
+
+        private_key = self.peer_relation.app_data.private_key
+        if not private_key:
+            return
+
+        csr = generate_csr(
+            private_key=private_key.encode("utf-8"),
+            subject=os.uname()[1],
+        ).strip()
+
+        self.certificates.request_certificate_creation(certificate_signing_request=csr)
+
+    def _on_certificate_available(self, event: CertificateAvailableEvent) -> None:
+        """Handle the on certificate available event."""
+        if not self.peer_relation.app_data:
+            event.defer()
+            return
+
+        self.write_file(content=self.peer_relation.app_data.private_key, path=KEY_FILE_PATH)
+        self.write_file(content=event.certificate, path=CERT_FILE_PATH)
+        self.write_file(content=event.ca, path=CA_FILE_PATH)
+
+    @staticmethod
+    def write_file(content: str, path: str) -> None:
+        """Write content in a file."""
+        os.makedirs(os.path.dirname(path), exist_ok=True)
+        with open(path, "w") as f:
+            f.write(content)
+
+    def _start_process_action(self, event: ActionEvent) -> None:
+        """Handle the start process action."""
+        if self.peer_relation.unit_data.pid:
+            event.fail(f"Process id {self.peer_relation.unit_data.pid} already running!")
+
+        app_type = self.config.app_type
+        logger.info(f"app: {app_type} : {type(app_type)}")
+        username = event.params["username"]
+        password = event.params["password"]
+        servers = event.params["servers"]
+        topic = event.params["topic_name"]
+        consumer_group_prefix = None
+        if "consumer_group_prefix" in event.params:
+            consumer_group_prefix = event.params["consumer_group_prefix"]
+
+        pid = self._start_process(
+            process_type=app_type,
+            username=username,
+            password=password,
+            servers=servers,
+            topic=topic,
+            consumer_group_prefix=consumer_group_prefix,
+            tls="disabled",
```

Comment:
shouldn't this be auto-inffered?

Reply:
I thought the for start or stop of producer/consumer we don't want relation with the kafka-app-test. The only way in which we can enable TLS is to relate it with the tls-certificates-operator. If you think this is a use case I can implement the logic to have TLS enabled while starting the producer/consumer with an action.

Reply:
I'm just saying that, exactly like you do for the relation, if the kafka-test-app is related to the tls operator, the process will use the encryption. Otherwise, we use the plain one. I mean, it is just to prevent the case when we relate tls to both kafka and kafka-test-app, and then we use tls disabled. If we base that on the relation, that should be self-driven. 

Anyhow, if this is too complex we can post pone this to a separate PR. It is not too critical